// @shopify/react-native-skia/lib/module/dom/nodes/datatypes/Gradient.js
import { TileMode } from "../../../skia/types";
import { enumKey } from "./Enum";
import { processTransformProps } from "./Transform";
export const transformOrigin = (origin, transform) => {
  "worklet";

  return [{
    translateX: origin.x
  }, {
    translateY: origin.y
  }, ...transform, {
    translateX: -origin.x
  }, {
    translateY: -origin.y
  }];
};
export const processGradientProps = (Skia, {
  colors,
  positions,
  mode,
  flags,
  ...transform
}) => {
  "worklet";

  const localMatrix = Skia.Matrix();
  processTransformProps(localMatrix, transform);
  return {
    colors: colors.map(color => Skia.Color(color)),
    positions: positions !== null && positions !== void 0 ? positions : null,
    mode: TileMode[enumKey(mode !== null && mode !== void 0 ? mode : "clamp")],
    flags,
    localMatrix
  };
};
export const getRect = (Skia, props) => {
  "worklet";

  const {
    x,
    y,
    width,
    height
  } = props;
  if (props.rect) {
    return props.rect;
  } else if (width !== undefined && height !== undefined) {
    return Skia.XYWHRect(x !== null && x !== void 0 ? x : 0, y !== null && y !== void 0 ? y : 0, width, height);
  } else {
    return undefined;
  }
};
